File: backend/app/core/confidence.py
"""
Confidence Score Calculator — 5 weighted dimensions:
  row_coverage (25%), data_completeness (25%), schema_match (20%),
  web_corroboration (15%), compliance_check (15%)
"""


def calculate_confidence(
    rows_used: int = 0,
    total_rows: int = 0,
    columns_used: list[str] | None = None,
    schema: list[dict] | None = None,
    question: str = "",
    web_results: list[dict] | None = None,
    sql_error: str | None = None,
    compliance_status: str = "compliant",
) -> dict:
    if not columns_used:
        columns_used = []
    if not schema:
        schema = []

    if sql_error:
        return {
            "score": 20,
            "level": "Low",
            "breakdown": {
                "row_coverage": 0,
                "data_completeness": 0,
                "schema_match": 20,
                "web_corroboration": 0,
                "compliance_check": 0,
            },
        }

    # 1. Row Coverage (25%)
    row_score = min(1.0, rows_used / total_rows) * 100 if total_rows > 0 else 0

    # 2. Data Completeness (25%)
    if columns_used and schema:
        schema_map = {col["name"].lower(): col for col in schema}
        missing_pcts = [schema_map.get(c.lower(), {}).get("missing_pct", 0) for c in columns_used]
        completeness_score = max(0, 100 - sum(missing_pcts) / max(len(missing_pcts), 1))
    else:
        completeness_score = 50

    # 3. Schema Match (20%)
    if columns_used:
        known_columns = {col["name"].lower() for col in schema}
        matched = sum(1 for c in columns_used if c.lower() in known_columns)
        schema_score = (matched / max(len(columns_used), 1)) * 100
    else:
        question_lower = question.lower()
        matches = sum(1 for col in schema if col["name"].lower() in question_lower)
        schema_score = min(100, matches * 25) if matches else 30

    # 4. Web Corroboration (15%)
    web_score = min(100, len(web_results) * 25) if web_results else 0

    # 5. Compliance Check (15%)
    compliance_score = {"compliant": 100, "warning": 50, "blocked": 0}.get(compliance_status, 100)

    total = (
        row_score * 0.25
        + completeness_score * 0.25
        + schema_score * 0.20
        + web_score * 0.15
        + compliance_score * 0.15
    )

    level = "High" if total >= 75 else ("Medium" if total >= 50 else "Low")

    return {
        "score": round(total),
        "level": level,
        "breakdown": {
            "row_coverage": round(row_score),
            "data_completeness": round(completeness_score),
            "schema_match": round(schema_score),
            "web_corroboration": round(web_score),
            "compliance_check": round(compliance_score),
        },
    }

File: backend-node/src/core/confidence.ts
/**
 * Confidence score calculator — mirrors Python backend/app/core/confidence.py exactly.
 */
import { Confidence, ColumnSchema, WebResult } from '../types';

export function calculateConfidence(opts: {
  rows_used?: number;
  total_rows?: number;
  columns_used?: string[];
  schema?: ColumnSchema[];
  question?: string;
  web_results?: WebResult[];
  sql_error?: string | null;
  compliance_status?: string;
}): Confidence {
  const {
    rows_used = 0, total_rows = 0, columns_used = [], schema = [],
    question = '', web_results = [], sql_error = null, compliance_status = 'compliant',
  } = opts;

  if (sql_error) {
    return {
      score: 20,
      level: 'Low',
      breakdown: { row_coverage: 0, data_completeness: 0, schema_match: 20, web_corroboration: 0, compliance_check: 0 },
    };
  }

  const rowScore = total_rows > 0 ? Math.min(1, rows_used / total_rows) * 100 : 0;

  let completenessScore: number;
  if (columns_used.length > 0 && schema.length > 0) {
    const schemaMap = new Map(schema.map((c) => [c.name.toLowerCase(), c]));
    const missingPcts = columns_used.map((c) => schemaMap.get(c.toLowerCase())?.missing_pct ?? 0);
    completenessScore = Math.max(0, 100 - missingPcts.reduce((a, b) => a + b, 0) / Math.max(missingPcts.length, 1));
  } else {
    completenessScore = 50;
  }

  let schemaScore: number;
  if (columns_used.length > 0) {
    const known = new Set(schema.map((c) => c.name.toLowerCase()));
    const matched = columns_used.filter((c) => known.has(c.toLowerCase())).length;
    schemaScore = (matched / Math.max(columns_used.length, 1)) * 100;
  } else {
    const qLower = question.toLowerCase();
    const matches = schema.filter((c) => qLower.includes(c.name.toLowerCase())).length;
    schemaScore = matches ? Math.min(100, matches * 25) : 30;
  }

  const webScore = Math.min(100, web_results.length * 25);
  const complianceScore: Record<string, number> = { compliant: 100, warning: 50, blocked: 0 };
  const compScore = complianceScore[compliance_status] ?? 100;

  const total = rowScore * 0.25 + completenessScore * 0.25 + schemaScore * 0.20 + webScore * 0.15 + compScore * 0.15;
  const level = total >= 75 ? 'High' : total >= 50 ? 'Medium' : 'Low';

  return {
    score: Math.round(total),
    level,
    breakdown: {
      row_coverage: Math.round(rowScore),
      data_completeness: Math.round(completenessScore),
      schema_match: Math.round(schemaScore),
      web_corroboration: Math.round(webScore),
      compliance_check: Math.round(compScore),
    },
  };
}
